cpp/arcticdb/column_store/memory_segment_impl.hpp

```diff
@@ -874,236 +784,19 @@ class SegmentInMemoryImpl {
-
-        for(const auto& column : columns_) {
-            output.columns_.push_back(std::make_shared<Column>(column->clone()));
-        }
-
-        output.string_pool_ = string_pool_->clone();
-        output.offset_ = offset_;
-        if(metadata_) {
-            google::protobuf::Any metadata;
-            metadata.CopyFrom(*metadata_);
-            output.metadata_ = std::make_unique<google::protobuf::Any>(std::move(metadata));
-        }
-        output.allow_sparse_ = allow_sparse_;
-        output.compacted_ = compacted_;
-        return output;
-    }
+    SegmentInMemoryImpl clone() const;
 
     void set_string_pool(const std::shared_ptr<StringPool>& string_pool) {
         string_pool_ = string_pool;
     }
 
-    // This implementation of split is incorrect Column::split doesn't work for sparse columns
-    // Will use filter_segment for now - which should do the same amount of copying as this implementation
-    std::vector<std::shared_ptr<SegmentInMemoryImpl>> old_split(size_t rows) {
-        std::vector<std::shared_ptr<SegmentInMemoryImpl>> output;
-        const auto output_size = std::ceil(double(row_count()) / double(rows));
-        for(auto i = 0u; i < output_size; ++i) {
-            output.push_back(std::make_shared<SegmentInMemoryImpl>());
-            if(has_string_pool())
-                (*output.rbegin())->set_string_pool(string_pool_);
-        }
-
-        for(const auto& column : folly::enumerate(columns_)) {
-            auto split_columns = Column::split(*column, rows);
-            for(auto split_col : folly::enumerate(split_columns)) {
-                output[split_col.index]->add_column(descriptor_->field(column.index), *split_col);
-            }
-        }
+    std::shared_ptr<SegmentInMemoryImpl> get_output_segment(size_t num_values, bool pre_allocate=true) const;
 
-        auto remaining_rows = row_count();
-        for(const auto& seg : output) {
-            const auto current_rows = std::min(remaining_rows, rows);
-            seg->set_row_data(static_cast<ssize_t>(current_rows) - 1);
-            remaining_rows -= current_rows;
-        }
-        util::check(remaining_rows == 0, "Didn't manage to set the rows correctly: {} rows left", remaining_rows);
-
-        return output;
-    }
-
-    std::shared_ptr<SegmentInMemoryImpl> get_output_segment(size_t num_values, bool pre_allocate=true) const {
-        std::shared_ptr<SegmentInMemoryImpl> output;
-        if (is_sparse()) {
-            output = allocate_sparse_segment(descriptor().id(), descriptor().index());
-        } else {
-            if (pre_allocate)
-                output = allocate_dense_segment(descriptor(), num_values);
-            else
-                output = allocate_dense_segment(StreamDescriptor(descriptor().id(), descriptor().index(), {}), num_values);
-        }
-        return output;
-    }
-
-    inline std::shared_ptr<SegmentInMemoryImpl> filter(const util::BitSet& filter_bitset,
```

Comment:
Do we or even should we `inline` those methods?

Reply:
I don't think we need to

Reply:
Anyway inline is mostly a hint that can be ignored. I don't believe it would impact the performance (it should be measured of course) but if we're concerned with this we can start running LTO for release builds.